src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -3656,22 +3656,182 @@ .maxstack  1
+
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, decls[0].Initializer!.Value,
+                expectedSyntax: "&M",
+                expectedType: null,
+                expectedConvertedType: "System.Void*",
+                expectedCandidateReason: CandidateReason.OverloadResolutionFailure,
+                expectedSymbolCandidates: new[] { "void C.M()" });
+
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, ((CastExpressionSyntax)decls[1].Initializer!.Value).Expression,
+                expectedSyntax: "&M",
+                expectedType: null,
+                expectedConvertedType: null,
+                expectedCandidateReason: CandidateReason.OverloadResolutionFailure,
+                expectedSymbolCandidates: new[] { "void C.M()" });
+        }
+
+        [Fact]
+        [WorkItem(44489, "https://github.com/dotnet/roslyn/issues/44489")]
+        public void AddressOf_ToDelegateType()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+using System;
+class C
+{
+    unsafe void M()
+    {
+        // This actually gets bound as a binary expression: (Action) & M
+        Action ptr1 = (Action)&M;
+        Action ptr2 = (Action)(&M);
+        Action ptr3 = &M;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (8,24): error CS0119: 'Action' is a type, which is not valid in the given context
+                //         Action ptr1 = (Action)&M;
+                Diagnostic(ErrorCode.ERR_BadSKunknown, "Action").WithArguments("System.Action", "type").WithLocation(8, 24),
+                // (8,24): error CS0119: 'Action' is a type, which is not valid in the given context
+                //         Action ptr1 = (Action)&M;
+                Diagnostic(ErrorCode.ERR_BadSKunknown, "Action").WithArguments("System.Action", "type").WithLocation(8, 24),
+                // (9,23): error CS8800: Cannot convert a & method group 'M' to delegate type 'M'.
+                //         Action ptr2 = (Action)(&M);
+                Diagnostic(ErrorCode.ERR_CannotConvertAddressOfToDelegate, "(Action)(&M)").WithArguments("M", "System.Action").WithLocation(9, 23),
+                // (10,23): error CS8800: Cannot convert a & method group 'M' to delegate type 'M'.
+                //         Action ptr3 = &M;
+                Diagnostic(ErrorCode.ERR_CannotConvertAddressOfToDelegate, "&M").WithArguments("M", "System.Action").WithLocation(10, 23)
+            );
+
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var decls = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().ToArray();
+
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, ((CastExpressionSyntax)decls[1].Initializer!.Value).Expression,
+                expectedSyntax: "(&M)",
+                expectedType: null,
+                expectedConvertedType: null,
```

Comment:
>null [](start = 39, length = 4)

The difference with the check below feels unexpected. Why does it matter whether there was an explicit cast in code?

Reply:
This mirrors how explicit conversions work in general: a `TypeInfo` does not contain information about explicit conversions. This mirrors how `(Action)M` would perform, though I do actually need to change the conversion to `Identity`, not `NoConversion`.

---
In reply to: [432042922](https://github.com/dotnet/roslyn/pull/44613#discussion_r432042922) [](ancestors = 432042922)